api4/ldap.go

```diff
@@ -38,3 +49,167 @@ func testLdap(c *Context, w http.ResponseWriter, r *http.Request) {
+		return
+	}
+
+	w.Write(b)
+}
+
+func linkLdapGroup(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.RequireRemoteId()
+	if c.Err != nil {
+		return
+	}
+
+	if !c.App.SessionHasPermissionTo(c.Session, model.PERMISSION_MANAGE_SYSTEM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	if c.App.License() == nil || !*c.App.License().Features.LDAPGroups {
+		c.Err = model.NewAppError("Api4.linkLdapGroup", "api.ldap.license.error", nil, "", http.StatusNotImplemented)
+		return
+	}
+
+	ldapGroup, err := c.App.GetLdapGroup(c.Params.RemoteId)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	if ldapGroup == nil {
+		c.Err = model.NewAppError("Api4.linkLdapGroup", "api.ldap.not_found", nil, "", http.StatusNotFound)
+		return
+	}
+
+	group, err := c.App.GetGroupByRemoteID(ldapGroup.RemoteId, model.GroupTypeLdap)
+	if err != nil && err.DetailedError != sql.ErrNoRows.Error() {
+		c.Err = err
+		return
+	}
+
+	var status int
+	var newOrUpdatedGroup *model.Group
+
+	// Group is already linked.
+	if group != nil {
+		if group.DeleteAt == 0 {
+			c.Err = model.NewAppError("Api4.linkLdapGroup", "api.ldap.already_linked", nil, "", http.StatusNotModified)
+			return
+		}
+
+		group.DeleteAt = 0
+		newOrUpdatedGroup, err = c.App.UpdateGroup(group)
+		if err != nil {
+			c.Err = err
+			return
+		}
+		status = http.StatusOK
+	} else {
+		// TODO: In a future phase of LDAP groups sync `Name` will be used for at-mentions and will be editable on
+		// the front-end so it will not have an initial value of `model.NewId()` but rather a slugified version of
+		// the LDAP group name with an appended duplicate-breaker.
+		newGroup := &model.Group{
+			Name:        model.NewId(),
+			DisplayName: ldapGroup.DisplayName,
+			RemoteId:    ldapGroup.RemoteId,
+			Type:        model.GroupTypeLdap,
+		}
+		newOrUpdatedGroup, err = c.App.CreateGroup(newGroup)
+		if err != nil {
+			c.Err = err
+			return
+		}
+		status = http.StatusCreated
+	}
+
+	b, marshalErr := json.Marshal(newOrUpdatedGroup)
+	if marshalErr != nil {
+		c.Err = model.NewAppError("Api4.linkLdapGroup", "api.ldap.marshal_error", nil, marshalErr.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(status)
+	w.Write(b)
+}
+
+func unlinkLdapGroup(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.RequireRemoteId()
+	if c.Err != nil {
+		return
+	}
+
+	if !c.App.SessionHasPermissionTo(c.Session, model.PERMISSION_MANAGE_SYSTEM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	if c.App.License() == nil || !*c.App.License().Features.LDAPGroups {
+		c.Err = model.NewAppError("Api4.unlinkLdapGroup", "api.ldap.license.error", nil, "", http.StatusNotImplemented)
+		return
+	}
+
+	group, err := c.App.GetGroupByRemoteID(c.Params.RemoteId, model.GroupTypeLdap)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	if group != nil && group.DeleteAt == 0 {
```

Comment:
I'm wondering if this and the next check could be a little more elegant if we deal with the `group == nil` case first? Or can we assume that if there's no `err` then the group is guaranteed to be not `nil` already?

Reply:
You're right it can be assumed not nil. Updated.